llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -3557,10 +3557,29 @@ LoadStoreOpcodes AIE2InstructionSelector::getLoadStoreOpcode(
                 /*OffsetOpcode=*/AIE2::VLDA_dmw_lda_am_ag_idx_imm};
       }
       if (RBID == AIE2::VRegBankID) {
-        FitsImmediateRange = checkImmediateRangeSplitting<11, 32, 32>(Offset);
-        return {/*ISelOpcode=*/AIE2::VLDA_dmw_lda_w_ag_idx_imm,
-                FitsImmediateRange,
-                /*OffsetOpcode=*/AIE2::VLDA_dmw_lda_w_ag_idx_imm};
+        unsigned OffsetOpcode;
+        // First try if the Instruction can be selected as multi-slot offset
+        // load
+        if (checkImmediateRangeSplitting<8, 32, 32>(Offset)) {
+          FitsImmediateRange = true;
+          ISelOpcode = OffsetOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
+        } else if (checkImmediateRange<8, 32>(Offset)) {
+          // When Offset is positive and one of the offset is in range of SlotB
+          ISelOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
+          OffsetOpcode = AIE2::VLDA_dmw_lda_w_ag_idx_imm;
+          FitsImmediateRange = true;
+        } else if (Offset.has_value() && (*Offset).isNegative() &&
+                   checkImmediateRange<8, 32>((*Offset) + 32)) {
+          // When Offset is negative and one of the offset is in range of SlotB
+          ISelOpcode = AIE2::VLDA_dmw_lda_w_ag_idx_imm;
+          OffsetOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
+          FitsImmediateRange = true;
+        } else {
+          FitsImmediateRange = checkImmediateRangeSplitting<11, 32, 32>(Offset);
```

Comment:
Nit: add a comment like `neither Offset nor Offset+32 fit for VLDB, try both for VLDA`